src/hashupdater.cpp

```diff
@@ -5,19 +5,47 @@
  */
 
 #include "hashupdater.hpp"
-#include "util.hpp"
+#include "util/threadpool.hpp"
 #include "world.hpp"
 
 PasswordHashUpdater::PasswordHashUpdater(Config& config, const std::unordered_map<HashFunc, std::shared_ptr<Hasher>>& passwordHashers)
     : _config(config)
     , _passwordHashers(passwordHashers)
-    , _database(nullptr)
-    , _terminating(false)
-    , _updateSem(0)
 {
-    this->_updateThread = std::thread([this] { this->updateThreadProc(); });
+}
+
+void PasswordHashUpdater::QueueUpdatePassword(const std::string& username, util::secure_string&& password, HashFunc hashFunc)
+{
+    auto updateThreadProc = [this](const void * state)
+    {
+        auto updateState = reinterpret_cast<const PasswordHashUpdater::UpdateState*>(state);
+
+        auto username = updateState->username;
+        auto updatedPassword = std::move(Hasher::SaltPassword(std::string(this->_config["PasswordSalt"]),
+                                                              username,
+                                                              std::move(const_cast<PasswordHashUpdater::UpdateState*>(updateState)->password)));
+        auto hashFunc = updateState->hashFunc;
+
+        if (hashFunc == NONE)
+            return;
+
+        updatedPassword = std::move(this->_passwordHashers[hashFunc]->hash(std::move(updatedPassword.str())));
+
+        this->CreateDbConnection()->Query("UPDATE `accounts` SET `password` = '$', `password_version` = # WHERE `username` = '$'",
+            updatedPassword.str().c_str(),
+            hashFunc,
+            username.c_str());
+
+        delete updateState;
+    };
 
-    auto dbType = util::lowercase(std::string(config["DBType"]));
+    auto state = reinterpret_cast<void*>(new UpdateState { username, std::move(password), hashFunc });
```

Comment:
This demonstrates how state can be used. I think having a `state` variable is potentially useful, but in reality it would probably suffice to just rely on lambda capturing since this implementation uses `std::function`.